File: discord-whatsapp-bridge/notification-bridge/watcher.lua
--[[
  Discord → WhatsApp notification watcher (Hammerspoon Lua module)

  Polls macOS Notification Center every N seconds for new Discord
  notifications and forwards them to a WhatsApp number via the Meta Cloud
  API. Reuses WHATSAPP_* credentials from the sibling .env file.

  Requires:
    - Hammerspoon (brew install --cask hammerspoon)
    - Full Disk Access granted to Hammerspoon
      (System Settings → Privacy & Security → Full Disk Access → +)
    - Discord desktop app running with native macOS notifications enabled
      for the channels you care about

  Usage from ~/.hammerspoon/init.lua:

      local watcher = dofile("/Users/you/.../notification-bridge/watcher.lua")
      watcher.start({
        env_path = "/Users/you/.../discord-whatsapp-bridge/.env",
        interval = 5,
      })
]]

local M = {}

local NOTIF_DB =
  os.getenv("HOME") ..
  "/Library/Group Containers/group.com.apple.usernoted/db2/db"
local DISCORD_BUNDLE = "com.hnc.Discord"
local STATE_DIR = os.getenv("HOME") .. "/.discord-wa-bridge"
local STATE_FILE = STATE_DIR .. "/state.json"

-- ── tiny helpers ──────────────────────────────────────────────────────────

-- Absolute paths so we don't depend on PATH (and so the login shell's
-- startup output doesn't get mixed into our results).
local SQLITE3 = "/usr/bin/sqlite3"
local PLUTIL  = "/usr/bin/plutil"

local function shell(cmd)
  -- hs.execute(cmd, false) → /bin/sh -c <cmd>, no .zshrc/.bash_profile.
  local out, ok = hs.execute(cmd, false)
  return out or "", ok
end

local function shellEscape(s)
  return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end

local function trim(s)
  return (s or ""):gsub("^%s+", ""):gsub("%s+$", "")
end

-- ── state ─────────────────────────────────────────────────────────────────

local function readState()
  local f = io.open(STATE_FILE, "r")
  if not f then return { last_rec_id = 0 } end
  local s = f:read("*a") or ""
  f:close()
  if s == "" then return { last_rec_id = 0 } end
  local ok, data = pcall(hs.json.decode, s)
  if ok and type(data) == "table" and tonumber(data.last_rec_id) then
    return { last_rec_id = tonumber(data.last_rec_id) }
  end
  return { last_rec_id = 0 }
end

local function writeState(state)
  shell("mkdir -p " .. shellEscape(STATE_DIR))
  local f = io.open(STATE_FILE, "w")
  if not f then return end
  f:write(hs.json.encode(state))
  f:close()
end

-- ── env loader ────────────────────────────────────────────────────────────

local function loadEnv(envPath)
  local env = {}
  local f = io.open(envPath, "r")
  if not f then return env end
  for line in f:lines() do
    if not line:match("^%s*#") then
      local k, v = line:match("^%s*([%w_]+)%s*=%s*(.-)%s*$")
      if k and v then env[k] = v end
    end
  end
  f:close()
  return env
end

-- ── notification DB reader ────────────────────────────────────────────────

local function discordAppId()
  local out, ok = shell(string.format(
    SQLITE3 .. " %s \"SELECT app_id FROM app WHERE identifier='%s';\" 2>&1",
    shellEscape(NOTIF_DB), DISCORD_BUNDLE
  ))
  if not ok or #out == 0 then return nil, out end
  if out:lower():match("unable to open") or out:lower():match("permission") then
    return nil, "Cannot read Notification Center DB — grant Full Disk Access to Hammerspoon."
  end
  local id = tonumber(out:match("(%-?%d+)"))
  return id, nil
end

local function fetchNewNotifications(lastRecId)
  local appId, err = discordAppId()
  if not appId then return {}, err end
  local sql = string.format(
    "SELECT rec_id FROM record WHERE app_id=%d AND rec_id>%d ORDER BY rec_id ASC;",
    appId, lastRecId
  )
  local out = shell(string.format(
    SQLITE3 .. " %s %s 2>/dev/null",
    shellEscape(NOTIF_DB), shellEscape(sql)
  ))
  local notifications = {}
  for recIdStr in out:gmatch("(%d+)") do
    local recId = tonumber(recIdStr)
    -- Dump the bplist blob to a temp file, convert to JSON via plutil.
    local tmp = string.format("/tmp/dwa-notif-%d.bplist", recId)
    shell(string.format(
      SQLITE3 .. " %s %s 2>/dev/null",
      shellEscape(NOTIF_DB),
      shellEscape(string.format("SELECT writefile('%s', data) FROM record WHERE rec_id=%d;", tmp, recId))
    ))
    local jsonStr = shell(string.format(PLUTIL .. " -convert json -o - %s 2>/dev/null", shellEscape(tmp)))
    shell("rm -f " .. shellEscape(tmp))
    if jsonStr and #jsonStr > 0 then
      local ok, payload = pcall(hs.json.decode, jsonStr)
      if ok and type(payload) == "table" then
        local req = payload.req or payload
        table.insert(notifications, {
          rec_id = recId,
          title = trim(req.titl or req.title or ""),
          subtitle = trim(req.subt or req.subtitle or ""),
          body = trim(req.body or req.Body or ""),
        })
      end
    end
  end
  return notifications, nil
end

-- ── WhatsApp senders ──────────────────────────────────────────────────────
-- Two backends: Meta Cloud API (needs Meta Business account) and Twilio
-- (needs a Twilio account but no Meta Business setup). The dispatcher picks
-- whichever set of env vars is fully populated, preferring Twilio.

local function urlencode(s)
  s = tostring(s or "")
  return (s:gsub("([^%w%-%._~])", function(c)
    return string.format("%%%02X", string.byte(c))
  end))
end

local function sendViaMeta(env, text)
  local url = string.format(
    "https://graph.facebook.com/%s/%s/messages",
    env.WHATSAPP_API_VERSION or "v20.0", env.WHATSAPP_PHONE_ID
  )
  local body = hs.json.encode({
    messaging_product = "whatsapp",
    to = env.WHATSAPP_TO,
    type = "text",
    text = { body = text:sub(1, 4096), preview_url = true },
  })
  hs.http.asyncPost(url, body, {
    ["Authorization"] = "Bearer " .. env.WHATSAPP_TOKEN,
    ["Content-Type"] = "application/json",
  }, function(status, respBody)
    if status < 200 or status >= 300 then
      print(string.format("[wa-bridge] meta send failed (%d): %s", status, tostring(respBody)))
    end
  end)
end

local function sendViaTwilio(env, text)
  local sid = env.TWILIO_ACCOUNT_SID
  local tok = env.TWILIO_AUTH_TOKEN
  local from = env.TWILIO_FROM   -- e.g. whatsapp:+14155238886 (sandbox number)
  local to   = env.TWILIO_TO     -- e.g. whatsapp:+15551234567
  -- Allow the user to omit the "whatsapp:" prefix and add it for them.
  if from and not from:match("^whatsapp:") then from = "whatsapp:" .. from end
  if to   and not to:match("^whatsapp:")   then to   = "whatsapp:" .. to   end
  local url = string.format(
    "https://api.twilio.com/2010-04-01/Accounts/%s/Messages.json", sid
  )
  local body = string.format(
    "From=%s&To=%s&Body=%s",
    urlencode(from), urlencode(to), urlencode(text:sub(1, 1500))
  )
  local auth = "Basic " .. hs.base64.encode(sid .. ":" .. tok)
  hs.http.asyncPost(url, body, {
    ["Authorization"] = auth,
    ["Content-Type"]  = "application/x-www-form-urlencoded",
  }, function(status, respBody)
    if status < 200 or status >= 300 then
      print(string.format("[wa-bridge] twilio send failed (%d): %s", status, tostring(respBody)))
    end
  end)
end

local function sendViaTelegram(env, text)
  local url = string.format("https://api.telegram.org/bot%s/sendMessage", env.TELEGRAM_BOT_TOKEN)
  local body = hs.json.encode({
    chat_id = env.TELEGRAM_CHAT_ID,
    text = text:sub(1, 4096),
    disable_web_page_preview = true,
  })
  hs.http.asyncPost(url, body, {
    ["Content-Type"] = "application/json",
  }, function(status, respBody)
    if status < 200 or status >= 300 then
      print(string.format("[wa-bridge] telegram send failed (%d): %s", status, tostring(respBody)))
    end
  end)
end

local function nonEmpty(s) return s and s ~= "" end

local function pickSender(env)
  if nonEmpty(env.TELEGRAM_BOT_TOKEN) and nonEmpty(env.TELEGRAM_CHAT_ID) then
    return "telegram", sendViaTelegram
  end
  if nonEmpty(env.TWILIO_ACCOUNT_SID) and nonEmpty(env.TWILIO_AUTH_TOKEN)
     and nonEmpty(env.TWILIO_FROM) and nonEmpty(env.TWILIO_TO) then
    return "twilio", sendViaTwilio
  end
  if nonEmpty(env.WHATSAPP_PHONE_ID) and nonEmpty(env.WHATSAPP_TOKEN)
     and nonEmpty(env.WHATSAPP_TO) then
    return "meta", sendViaMeta
  end
  return nil, nil
end

local function sendMessage(env, text)
  local provider, fn = pickSender(env)
  if not fn then
    print("[wa-bridge] no provider configured (need TELEGRAM_*, TWILIO_*, or WHATSAPP_* in .env)")
    return
  end
  fn(env, text)
end

-- ── filter + format ───────────────────────────────────────────────────────

local function parseFilter(env)
  local raw = env.DISCORD_CHANNEL_FILTER
  if not raw or raw == "" then return nil end
  local out = {}
  for s in raw:gmatch("([^,]+)") do
    local v = trim(s):lower()
    if v ~= "" then table.insert(out, v) end
  end
  return out
end

local function passesFilter(n, filter)
  if not filter then return true end
  local hay = (n.title .. " " .. n.subtitle):lower()
  for _, s in ipairs(filter) do
    if hay:find(s, 1, true) then return true end
  end
  return false
end

local function formatText(n)
  local parts = {}
  if n.title ~= "" then table.insert(parts, "[" .. n.title .. "]") end
  if n.subtitle ~= "" then table.insert(parts, n.subtitle .. ":") end
  if n.body ~= "" then table.insert(parts, n.body) end
  local s = table.concat(parts, " ")
  if s == "" then return "(empty Discord notification)" end
  return s
end

-- ── main poll ─────────────────────────────────────────────────────────────

local function pollOnce(opts)
  local env = loadEnv(opts.env_path)
  local provider = pickSender(env)
  if not provider then
    print("[wa-bridge] no provider configured — set either TWILIO_* (ACCOUNT_SID, AUTH_TOKEN, FROM, TO) or WHATSAPP_* (PHONE_ID, TOKEN, TO) in .env")
    return
  end
  local state = readState()
  local notifications, err = fetchNewNotifications(state.last_rec_id)
  if err then
    print("[wa-bridge] " .. err)
    return
  end
  if #notifications == 0 then return end
  local filter = parseFilter(env)
  local sent = 0
  for _, n in ipairs(notifications) do
    if passesFilter(n, filter) then
      sendMessage(env, formatText(n))
      sent = sent + 1
    end
    state.last_rec_id = n.rec_id
  end
  writeState(state)
  if sent > 0 then
    print(string.format("[wa-bridge] forwarded %d (last rec_id=%d)", sent, state.last_rec_id))
  end
end

-- ── public API ────────────────────────────────────────────────────────────

function M.start(opts)
  opts = opts or {}
  if not opts.env_path then
    error("watcher.start: env_path is required")
  end
  local interval = tonumber(opts.interval) or 5
  -- On first start, sweep current rec_id forward so we don't replay old
  -- notifications stored in the DB. Comment this out if you want a backfill.
  do
    local state = readState()
    if state.last_rec_id == 0 then
      local out = shell(string.format(
        SQLITE3 .. " %s %s 2>/dev/null",
        shellEscape(NOTIF_DB),
        shellEscape("SELECT COALESCE(MAX(rec_id), 0) FROM record;")
      ))
      local maxId = tonumber(out:match("(%d+)"))
      if maxId then
        writeState({ last_rec_id = maxId })
        print(string.format("[wa-bridge] initialized at rec_id=%d (no backfill)", maxId))
      end
    end
  end
  M._timer = hs.timer.doEvery(interval, function() pollOnce(opts) end)
  print(string.format("[wa-bridge] started — polling every %ds", interval))
end

function M.stop()
  if M._timer then M._timer:stop(); M._timer = nil end
  print("[wa-bridge] stopped")
end

function M.testOnce(opts)
  pollOnce(opts or {})
end

-- Send a manual smoke test message to verify WhatsApp credentials.
function M.smoke(opts)
  opts = opts or {}
  local env = loadEnv(opts.env_path)
  sendMessage(env, "✅ Hammerspoon watcher smoke test — if you see this, .env credentials are good.")
  print("[wa-bridge] smoke test sent")
end

return M

File: discord-whatsapp-bridge/python-bridge/watcher.py
#!/usr/bin/env python3
"""
discord-watcher — Forward macOS Discord notifications to Telegram (or Twilio
WhatsApp, or Meta WhatsApp Cloud). Pure-Python replacement for the
Hammerspoon-based watcher.lua.

Polls ~/Library/Group Containers/group.com.apple.usernoted/db2/db every N
seconds, filters for Discord (com.hnc.Discord) notifications, and forwards
new ones using whichever provider is configured in .env.

Run:
    python3 watcher.py            # poll forever
    python3 watcher.py --once     # one poll, exit (useful for cron/testing)
    python3 watcher.py --smoke    # send a test message via the configured provider

The launchd plist installs this as a LaunchAgent that runs continuously.

Full Disk Access must be granted to whichever Python binary runs this
(System Settings → Privacy & Security → Full Disk Access). Without it the
SQLite open will fail with "unable to open database file".
"""

from __future__ import annotations

import argparse
import json
import os
import plistlib
import sqlite3
import sys
import time
from pathlib import Path
from urllib import error, request
from urllib.parse import urlencode

NOTIF_DB = Path.home() / "Library/Group Containers/group.com.apple.usernoted/db2/db"
# Discord actually registers its bundle as lowercase 'discord' on current
# macOS versions. We accept either to be future-proof.
DISCORD_BUNDLES = ("com.hnc.discord", "com.hnc.Discord")
STATE_DIR = Path.home() / ".discord-wa-bridge"
STATE_FILE = STATE_DIR / "state.json"


def log(msg: str) -> None:
    print(f"[watcher] {msg}", flush=True)


# ── .env loader ────────────────────────────────────────────────────────────

def load_env(env_path: Path) -> dict:
    env: dict = {}
    if not env_path.exists():
        return env
    for raw in env_path.read_text().splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, _, v = line.partition("=")
        env[k.strip()] = v.strip()
    return env


# ── state file ─────────────────────────────────────────────────────────────

def load_state() -> dict:
    if STATE_FILE.exists():
        try:
            data = json.loads(STATE_FILE.read_text())
            if isinstance(data, dict) and isinstance(data.get("last_rec_id"), int):
                return data
        except Exception:
            pass
    return {"last_rec_id": 0}


def save_state(state: dict) -> None:
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    STATE_FILE.write_text(json.dumps(state))


# ── Notification Center DB reader ──────────────────────────────────────────

def _open_db() -> sqlite3.Connection:
    return sqlite3.connect(f"file:{NOTIF_DB}?mode=ro", uri=True, timeout=2)


def fetch_new_discord(last_rec_id: int):
    if not NOTIF_DB.exists():
        return [], f"Notification DB missing at {NOTIF_DB}"
    try:
        conn = _open_db()
    except sqlite3.OperationalError as e:
        return [], f"Cannot open Notification DB: {e}. Grant Full Disk Access to Python."
    try:
        cur = conn.cursor()
        placeholders = ",".join("?" * len(DISCORD_BUNDLES))
        cur.execute(
            f"SELECT app_id FROM app WHERE identifier IN ({placeholders})",
            DISCORD_BUNDLES,
        )
        app_ids = [r[0] for r in cur.fetchall()]
        if not app_ids:
            return [], None  # Discord hasn't posted any notifications yet
        ph = ",".join("?" * len(app_ids))
        cur.execute(
            f"SELECT rec_id, data FROM record WHERE app_id IN ({ph}) AND rec_id > ? "
            f"ORDER BY rec_id ASC",
            (*app_ids, last_rec_id),
        )
        rows = cur.fetchall()
    finally:
        conn.close()

    out = []
    for rec_id, blob in rows:
        try:
            payload = plistlib.loads(blob)
        except Exception as e:
            log(f"skip rec {rec_id}: bplist parse failed ({e})")
            continue
        # The notification text lives under 'req' on recent macOS versions, or
        # at top-level on older ones. Fall back gracefully.
        req = payload.get("req") if isinstance(payload, dict) else None
        if not isinstance(req, dict):
            req = payload if isinstance(payload, dict) else {}
        out.append({
            "rec_id": rec_id,
            "title":    (req.get("titl") or req.get("title") or "").strip(),
            "subtitle": (req.get("subt") or req.get("subtitle") or "").strip(),
            "body":     (req.get("body") or req.get("Body") or "").strip(),
            # Hold the raw payload so the OCR-enrich path can pull out the
            # discord:// deep link to navigate Discord to this exact message.
            "_payload": payload,
        })
    return out, None


def get_max_rec_id():
    if not NOTIF_DB.exists():
        return None
    try:
        conn = _open_db()
    except sqlite3.OperationalError:
        return None
    try:
        cur = conn.cursor()
        cur.execute("SELECT COALESCE(MAX(rec_id), 0) FROM record")
        row = cur.fetchone()
        return int(row[0]) if row and row[0] is not None else 0
    finally:
        conn.close()


# ── HTTP helpers ───────────────────────────────────────────────────────────

def _post(url: str, body: bytes, headers: dict):
    req = request.Request(url, data=body, method="POST", headers=headers)
    try:
        with request.urlopen(req, timeout=10) as resp:
            return resp.status, resp.read().decode("utf-8", "replace")
    except error.HTTPError as e:
        return e.code, e.read().decode("utf-8", "replace")
    except Exception as e:
        return 0, str(e)


def _post_json(url: str, payload: dict, headers: dict | None = None):
    h = {"Content-Type": "application/json"}
    if headers:
        h.update(headers)
    return _post(url, json.dumps(payload).encode(), h)


def _post_form(url: str, form: dict, headers: dict | None = None):
    h = {"Content-Type": "application/x-www-form-urlencoded"}
    if headers:
        h.update(headers)
    return _post(url, urlencode(form).encode(), h)


# ── providers ──────────────────────────────────────────────────────────────

def send_telegram(env: dict, text: str) -> None:
    token = env["TELEGRAM_BOT_TOKEN"]
    chat = env["TELEGRAM_CHAT_ID"]
    chat_val = int(chat) if chat.lstrip("-").isdigit() else chat
    url = f"https://api.telegram.org/bot{token}/sendMessage"
    code, body = _post_json(url, {
        "chat_id": chat_val,
        "text": text[:4096],
        "disable_web_page_preview": True,
    })
    if not (200 <= code < 300):
        log(f"telegram send failed {code}: {body[:200]}")


def send_twilio(env: dict, text: str) -> None:
    import base64
    sid = env["TWILIO_ACCOUNT_SID"]
    tok = env["TWILIO_AUTH_TOKEN"]
    from_ = env["TWILIO_FROM"]
    to = env["TWILIO_TO"]
    if not from_.startswith("whatsapp:"): from_ = "whatsapp:" + from_
    if not to.startswith("whatsapp:"): to = "whatsapp:" + to
    url = f"https://api.twilio.com/2010-04-01/Accounts/{sid}/Messages.json"
    auth = base64.b64encode(f"{sid}:{tok}".encode()).decode()
    code, body = _post_form(
        url, {"From": from_, "To": to, "Body": text[:1500]},
        headers={"Authorization": f"Basic {auth}"},
    )
    if not (200 <= code < 300):
        log(f"twilio send failed {code}: {body[:200]}")


def send_meta(env: dict, text: str) -> None:
    phone_id = env["WHATSAPP_PHONE_ID"]
    token = env["WHATSAPP_TOKEN"]
    to = env["WHATSAPP_TO"]
    api_v = env.get("WHATSAPP_API_VERSION", "v20.0")
    url = f"https://graph.facebook.com/{api_v}/{phone_id}/messages"
    code, body = _post_json(url, {
        "messaging_product": "whatsapp",
        "to": to,
        "type": "text",
        "text": {"body": text[:4096], "preview_url": True},
    }, headers={"Authorization": f"Bearer {token}"})
    if not (200 <= code < 300):
        log(f"meta send failed {code}: {body[:200]}")


def pick_sender(env: dict):
    if env.get("TELEGRAM_BOT_TOKEN") and env.get("TELEGRAM_CHAT_ID"):
        return "telegram", send_telegram
    if (env.get("TWILIO_ACCOUNT_SID") and env.get("TWILIO_AUTH_TOKEN")
            and env.get("TWILIO_FROM") and env.get("TWILIO_TO")):
        return "twilio", send_twilio
    if (env.get("WHATSAPP_PHONE_ID") and env.get("WHATSAPP_TOKEN")
            and env.get("WHATSAPP_TO")):
        return "meta", send_meta
    return None, None


# ── filter + format ───────────────────────────────────────────────────────

def parse_filter(env: dict):
    raw = env.get("DISCORD_CHANNEL_FILTER", "")
    return [s.strip().lower() for s in raw.split(",") if s.strip()]


def passes_filter(n: dict, filt) -> bool:
    if not filt:
        return True
    hay = f"{n['title']} {n['subtitle']}".lower()
    return any(s in hay for s in filt)


def format_text(n: dict) -> str:
    parts = []
    if n["title"]:    parts.append(f"[{n['title']}]")
    if n["subtitle"]: parts.append(f"{n['subtitle']}:")
    if n["body"]:     parts.append(n["body"])
    return " ".join(parts) or "(empty Discord notification)"


# ── optional OCR enrichment ────────────────────────────────────────────────
# When the notification body is very short (likely a bot post with embeds
# that macOS notifications don't surface), open the Discord client to the
# specific message, screenshot, and OCR. Adds 3-5s latency per enriched
# message and steals focus. Enable with OCR_ENRICH=true in .env.

def maybe_enrich(env: dict, n: dict, base_text: str) -> str:
    if (env.get("OCR_ENRICH", "").lower() not in ("1", "true", "yes", "on")):
        return base_text
    body = n.get("body", "")
    threshold = int(env.get("OCR_ENRICH_BODY_THRESHOLD", "30"))
    if len(body) >= threshold:
        return base_text  # plain-text message, no enrichment needed
    payload = n.get("_payload")
    if not payload:
        return base_text
    try:
        # Lazy import so missing PyObjC bindings don't break the watcher.
        from ocr_enrich import enrich
    except Exception as e:
        log(f"OCR_ENRICH set but ocr_enrich import failed: {e}")
        return base_text
    try:
        ocr_text = enrich(payload, wait_seconds=float(env.get("OCR_WAIT", "2.5")))
    except Exception as e:
        log(f"enrich error: {e}")
        return base_text
    if not ocr_text or not ocr_text.strip():
        return base_text
    # Trim to a reasonable size — Telegram max is 4096
    snippet = ocr_text.strip()[:2500]
    return f"{base_text}\n\n── via OCR ──\n{snippet}"


# ── main loop ─────────────────────────────────────────────────────────────

def poll_once(env_path: Path) -> int:
    env = load_env(env_path)
    provider, fn = pick_sender(env)
    if not fn:
        log("no provider configured — need TELEGRAM_*, TWILIO_*, or WHATSAPP_* in .env")
        return 0
    state = load_state()
    notifications, err = fetch_new_discord(state["last_rec_id"])
    if err:
        log(err)
        return 0
    if not notifications:
        return 0
    filt = parse_filter(env)
    sent = 0
    for n in notifications:
        if passes_filter(n, filt):
            try:
                text = format_text(n)
                text = maybe_enrich(env, n, text)
                fn(env, text)
                sent += 1
            except Exception as e:
                log(f"send error rec {n['rec_id']}: {e}")
        state["last_rec_id"] = n["rec_id"]
    save_state(state)
    if sent:
        log(f"forwarded {sent} (last rec_id={state['last_rec_id']})")
    return sent


def smoke(env_path: Path) -> None:
    env = load_env(env_path)
    provider, fn = pick_sender(env)
    if not fn:
        log("no provider configured in .env")
        return
    try:
        fn(env, "✅ Python watcher smoke test — if you see this, .env credentials are good.")
        log(f"smoke test sent via {provider}")
    except Exception as e:
        log(f"smoke send failed: {e}")


def initialize_state_if_empty():
    state = load_state()
    if state["last_rec_id"] == 0:
        max_id = get_max_rec_id()
        if max_id is not None:
            save_state({"last_rec_id": max_id})
            log(f"initialized at rec_id={max_id} (no backfill)")


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument(
        "--env",
        default=str(Path(__file__).resolve().parent.parent / ".env"),
        help="Path to .env file (default: ../.env relative to this script)",
    )
    ap.add_argument("--interval", type=int, default=5,
                    help="Seconds between polls (default: 5)")
    ap.add_argument("--once", action="store_true",
                    help="Poll once and exit")
    ap.add_argument("--smoke", action="store_true",
                    help="Send a smoke-test message via the configured provider and exit")
    args = ap.parse_args()
    env_path = Path(args.env)

    if args.smoke:
        smoke(env_path)
        return
    if args.once:
        poll_once(env_path)
        return

    initialize_state_if_empty()
    log(f"started — polling every {args.interval}s, env={env_path}")
    while True:
        try:
            poll_once(env_path)
        except KeyboardInterrupt:
            log("interrupted, exiting")
            return
        except Exception as e:
            log(f"poll error: {e}")
        time.sleep(args.interval)


if __name__ == "__main__":
    main()
